optaplanner-core/src/main/java/org/optaplanner/core/impl/score/stream/drools/common/DroolsUniAccumulateFunctionAdapter.java

```diff
@@ -0,0 +1,131 @@
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.kie.api.runtime.rule.AccumulateFunction;
+import org.optaplanner.core.api.score.stream.uni.UniConstraintCollector;
+import org.optaplanner.core.impl.score.stream.drools.uni.UniConstraintCollectorAdapter;
+
+/**
+ * Drools {@link AccumulateFunction} that calls {@link UniConstraintCollector} underneath.
+ * @param <A> input to accumulate
+ * @param <ResultContainer_> implementation detail
+ * @param <NewA> result of accumulation
+ */
+public final class DroolsUniAccumulateFunctionAdapter<A, ResultContainer_ extends Serializable, NewA>
+        implements AccumulateFunction<ResultContainer_> {
+
+    private final Supplier<ResultContainer_> supplier;
+    private final BiFunction<ResultContainer_, A, Runnable> accumulator;
+    private final Function<ResultContainer_, NewA> finisher;
+    private final Map<Object, Undo> undoMap = new HashMap<>(0);
+
+    public DroolsUniAccumulateFunctionAdapter(UniConstraintCollectorAdapter<A, ResultContainer_, NewA> collector) {
+        this.supplier = collector.get().supplier();
+        this.accumulator = collector.get().accumulator();
+        this.finisher = collector.get().finisher();
+    }
+
+    @Override
+    public ResultContainer_ createContext() {
+        return supplier.get();
+    }
+
+    @Override
+    public void init(ResultContainer_ context) {
+        undoMap.clear();
+    }
+
+    @Override
+    public void accumulate(ResultContainer_ context, Object value) {
+        undoMap.compute(value, (k, v) -> {
+           if (v == null) {
+               Runnable runnable = accumulator.apply(context, (A) value);
+               return new Undo(runnable);
+           } else {
+               v.incrementUseCount();
+               return v;
+           }
+        });
+    }
+
+    @Override
+    public void reverse(ResultContainer_ context, Object value) {
+        Undo undo = undoMap.get(value);
+        if (undo.decrementUseCountAndGet() == 0) { // Don't keep empty counters
+            undoMap.remove(value);
+        }
+        undo.run();
+    }
+
+    @Override
+    public Object getResult(ResultContainer_ context) {
+        return finisher.apply(context);
+    }
+
+    @Override
+    public boolean supportsReverse() {
+        return true;
+    }
+
+    @Override
+    public Class<?> getResultType() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void writeExternal(ObjectOutput out) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public void readExternal(ObjectInput in) {
+        throw new UnsupportedOperationException();
+    }
+
+    private static final class Undo {
+
+        private final Runnable runnable;
+        private final AtomicLong useCount = new AtomicLong(1);
```

Comment:
AtomicLong do synchronization (so light locking). But this is used in a single-threaded context, so that not really needed, performance wise. How about adding a comment (maybe a TODO?) to point that out?